src/cuda_eval.cpp

```diff
@@ -408,6 +408,26 @@ static void jitc_cuda_render_var(uint32_t index, const Variable *v) {
              *a2 = v->dep[2] ? jitc_var(v->dep[2]) : nullptr,
              *a3 = v->dep[3] ? jitc_var(v->dep[3]) : nullptr;
 
+    bool f32_upcast = jitc_is_half(v) && var_kind_fp16_unsupported_cuda[v->kind];
+
+    if (f32_upcast) {
+        Variable* b = const_cast<Variable*>(v);
+        b->type = (uint32_t)VarType::Float32;
+        for (size_t i = 0; i < 4; ++i) {
+            Variable* dep = b->dep[i] ? jitc_var(b->dep[i]) : nullptr;
```

Comment:
(Despite the comment below). Would it not be easier to understand to merge these two loops? You can add a check to see if the variable has already been set to `Float32`. I think the current implementation may generate the `cvt.f32.f16` instruction multiple times for operations that receive the same operand twice.